### PR TITLE
IC-1646: Implement access restrictions on for getSentReferral

### PR DESCRIPTION
**Pull request by andrewburrowsmoj**
## What does this pull request do?

Adds Authorization checks for getSentReferrals.
This is a simple auth check just on service provider.
Once approved this will be expanded to other end points and for subcontractor checks.

## What is the intent behind these changes?

Prevent access to referrals if the user is not part of the correct service provider organization.

### CODE REVIEW

**Review by rgforsyth — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt [hunk 1]
```diff
@@ -27,31 +27,29 @@ import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Cancell
...
-      if (it.id != hmppsAuthGroupId) {
-        logger.error("AuthUser [id=${authUser.id}] does not have access to referral [id=$referralId]")
-        throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "unauthorized to access referral [id=$referralId]")
-      }
+    val authUser = userMapper.fromToken(authentication)
+    val referral = referralService.getReferral(referralId)
+      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "referral not found [id=$referralId]")
+    if(!referralAccessChecker.forServiceProviderUser(referral, authUser)){
+      logger.error("AuthUser [id=${authUser.id}] does not have access to referral [id=$referralId]")
+      throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "unauthorized to access referral [id=$referralId]")
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:52]: small thing - http 401 has a misleading name. it's actually to signal the request is lacking authentication, not authorization. the appropriate response here is 403 forbidden.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt [hunk 2]
```diff
@@ -27,31 +27,29 @@ import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Cancell
...
-    val hmppsAuthGroupId = hmppsAuthService.getServiceProviderOrganizationForUser(authUser)
-    val serviceProvider = serviceProviderService.getServiceProviderByReferralId(referralId)
-      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
-    serviceProvider.let {
-      if (it.id != hmppsAuthGroupId) {
-        logger.error("AuthUser [id=${authUser.id}] does not have access to referral [id=$referralId]")
-        throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "unauthorized to access referral [id=$referralId]")
-      }
+    val authUser = userMapper.fromToken(authentication)
+    val referral = referralService.getReferral(referralId)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:48]: this is fine for now, but later let's refactor the service layer to hide this complexity from the controller.

that way any other controller (not just referral controllers) that want to get a referral by ID end up having the access checks applied too.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt (hunk 1 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt]: Thanks. I always get mixed up with those two definitions. I've fixed it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DynamicFrameworkContractService.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ServiceProvider
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ReferralRepository
+import java.util.UUID
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.DynamicFrameworkContract
+
+@Service
+class DynamicFrameworkContractService(val referralRepository: ReferralRepository) {
+  fun getDynamicFrameworkContractByReferralId(id: UUID): DynamicFrameworkContract? {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/DynamicFrameworkContractService.kt:12]: just a note: this is currently unused

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ServiceProviderService.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service
+
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ServiceProvider
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ReferralRepository
+import java.util.UUID
+
+@Service
+class ServiceProviderService(val referralRepository: ReferralRepository) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ServiceProviderService.kt:10]: note: this is also unused

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/ReferralAccessTest.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ServiceProviderRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.HMPPSAuthService
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.AuthUserFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.DynamicFrameworkContractFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.InterventionFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.JwtTokenFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.ReferralFactory
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.util.ServiceProviderFactory
+
+@Import(MockAuthContext::class)
```
Comment (tomsmyers) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/ReferralAccessTest.kt:41]: nit: can't we use a simple `MockBean` field in this class to achieve this?

Comment (tomsmyers) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/ReferralAccessTest.kt:41]: oh i see we do this lower down anyway 🤔

Comment (tomsmyers) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/integration/ReferralAccessTest.kt]: i checked this without this import and it does still work. will remove it.

**Review by tomsmyers — APPROVED** [review]
approving but noting the comments around refactoring the access check in ReferralController

**Pull request merged**